pkg/controller/pravegacluster/upgrade.go

```diff
@@ -0,0 +1,372 @@
+	}
+
+	targetImage := fmt.Sprintf("%s:%s", p.Spec.Pravega.ImageRepository, p.Status.TargetVersion)
+
+	if sts.Spec.Template.Spec.Containers[0].Image != targetImage {
+		// Need to update pod template
+		// This will trigger the rolling upgrade process
+		log.Printf("updating statefulset (%s) template image to '%s'", sts.Name, targetImage)
+
+		sts.Spec.Template = pravega.MakeSegmentStorePodTemplate(p)
+		//sts.Spec.Template.Spec.Containers[0].Image = targetImage
+		err = r.client.Update(context.TODO(), sts)
+		if err != nil {
+			return false, err
+		}
+		// Updated pod template. Upgrade process has been triggered
+		return false, nil
+	}
+
+	// Pod template already updated
+
+	log.Printf("statefulset (%s) status: %d updated, %d ready, %d target", sts.Name,
+		sts.Status.UpdatedReplicas, sts.Status.ReadyReplicas, sts.Status.Replicas)
+
+	// Check whether the upgrade is in progress or has completed
+	if sts.Status.UpdatedReplicas != sts.Status.Replicas ||
+		sts.Status.UpdatedReplicas != sts.Status.ReadyReplicas {
+		// Upgrade still in progress
+		return false, nil
+	}
+
+	// StatefulSet upgrade completed
+	return true, nil
+}
+
+func (r *ReconcilePravegaCluster) syncBookkeeperVersion(p *pravegav1alpha1.PravegaCluster) (synced bool, err error) {
+	sts := &appsv1.StatefulSet{}
+	name := util.StatefulSetNameForBookie(p.Name)
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: p.Namespace}, sts)
+	if err != nil {
+		return false, fmt.Errorf("failed to get statefulset (%s): %v", sts.Name, err)
+	}
+
+	targetImage := fmt.Sprintf("%s:%s", p.Spec.Bookkeeper.ImageRepository, p.Status.TargetVersion)
+
+	if sts.Spec.Template.Spec.Containers[0].Image != targetImage {
+		// Need to update pod template
+		// This will trigger the rolling upgrade process
+		log.Printf("updating statefulset (%s) template image to '%s'", sts.Name, targetImage)
+		sts.Spec.Template = pravega.MakeBookiePodTemplate(p)
+		// sts.Spec.Template.Spec.Containers[0].Image = targetImage
+		// sts.Spec.Template.SetAnnotations(map[string]string{"pravega.version": p.Status.TargetVersion})
+		err = r.client.Update(context.TODO(), sts)
+		if err != nil {
+			return false, err
+		}
+		// Updated pod template
+		return false, nil
+	}
+
+	// Pod template already updated
+
+	log.Printf("statefulset (%s) status: %d updated, %d ready, %d target", sts.Name,
+		sts.Status.UpdatedReplicas, sts.Status.ReadyReplicas, sts.Status.Replicas)
+
+	// Check whether the upgrade is in progress or has completed
+	if sts.Status.UpdatedReplicas == sts.Status.Replicas &&
+		sts.Status.UpdatedReplicas == sts.Status.ReadyReplicas {
+		// StatefulSet upgrade completed
+		// TODO: wait until there is no under replicated ledger
+		// https://bookkeeper.apache.org/docs/4.7.2/reference/cli/#listunderreplicated
+		return true, nil
+	}
+
+	// Upgrade still in progress
+	// If all replicas are ready, upgrade an old pod
```

Comment:
I am wondering why Bookkeeper upgrading is using `onDelete` method instead of `rollingUpgrade` in SegmentStore?

Reply:
Good question. I'll create a document explaining the upgrade process and the reason behind those decisions.